File: ts/src/http.ts
/**
 * Internal HTTP client. Wraps fetch and translates non-2xx responses + network
 * failures into RealmError. Server error envelopes are expected in the shape:
 *   { error: { code, message }, ...siblings }
 * where siblings (e.g. mfa_challenge_token) flow into RealmError.details.
 *
 * Authorization is sourced from a platform-token manager (SPEC §4.0):
 * every outbound request — including `auth.login` — uses the
 * short-lived platform token. The raw API key only travels on the
 * `POST /auth/platform-token` call inside `PlatformTokenManager`.
 */

import { RealmError, statusToCode, isKnownCode, type ErrorCode } from "./errors.js";
import type { Logger } from "./logger.js";
import { NOOP_LOGGER, redactCredential } from "./logger.js";
import type { PlatformTokenManager } from "./platform-token-manager.js";

export interface HttpClientOptions {
  baseUrl: string;
  fetch?: typeof fetch;
  /**
   * Source of the platform token used for normal management/auth calls.
   * When omitted, requests go out unauthenticated (e.g. JWKS, public
   * info routes). Per-call `bearer` always wins.
   */
  platformTokens?: PlatformTokenManager;
  logger?: Logger;
  /**
   * The end user's *verified* access JWT, forwarded as `X-User-Token` on every
   * request this client makes (SPEC §4 verified on-behalf-of; ADR-056). Set by
   * `realm.withUserToken(jwt)` rather than directly: a client carrying a user
   * token is a DERIVED client, so a user's identity can never leak onto the
   * long-lived realm handle. A per-call `headers["x-user-token"]` still wins.
   */
  userToken?: string;
}

export interface RequestOptions {
  method?: string;
  path: string;
  query?: Record<string, string | number | undefined>;
  body?: unknown;
  /** Per-call bearer override (e.g. user JWT for sessions endpoints). */
  bearer?: string;
  /** Extra headers (e.g. Origin) merged after defaults. */
  headers?: Record<string, string>;
  /**
   * Skip the platform-token mint for this call (used internally so the
   * manager itself doesn't recurse). Defaults to false.
   */
  skipPlatformToken?: boolean;
}

export class HttpClient {
  private readonly baseUrl: string;
  private readonly fetch: typeof fetch;
  private readonly platformTokens?: PlatformTokenManager;
  private readonly logger: Logger;
  private readonly opts: HttpClientOptions;
  private readonly userToken?: string;

  constructor(opts: HttpClientOptions) {
    this.opts = opts;
    this.baseUrl = opts.baseUrl.replace(/\/+$/, "");
    this.fetch = opts.fetch ?? globalThis.fetch.bind(globalThis);
    this.platformTokens = opts.platformTokens;
    this.logger = opts.logger ?? NOOP_LOGGER;
    this.userToken = opts.userToken;
  }

  /**
   * Returns a COPY of this client that forwards `accessJWT` as `X-User-Token`.
   * The platform-token manager is shared, so the derived client reuses the
   * cached platform token instead of re-minting one per user.
   */
  withUserToken(accessJWT: string): HttpClient {
    return new HttpClient({ ...this.opts, userToken: accessJWT });
  }

  async request<T = unknown>(opts: RequestOptions): Promise<T> {
    const url = this.buildUrl(opts.path, opts.query);
    const headers: Record<string, string> = {
      "content-type": "application/json",
      "accept": "application/json",
    };

    let bearer = opts.bearer;
    if (!bearer && !opts.skipPlatformToken && this.platformTokens) {
      bearer = await this.platformTokens.getToken();
    }
    if (bearer) headers["authorization"] = `Bearer ${bearer}`;
    // The client-scoped user token is ADDITIVE — the platform token stays the
    // wire bearer (ADR-056). Set before the per-call merge so an explicit
    // header still wins. Per-call keys are lower-cased on the way in: `fetch`
    // treats header names case-insensitively and would otherwise APPEND a
    // differently-cased duplicate ("a, b") instead of overriding it.
    if (this.userToken) headers["x-user-token"] = this.userToken;
    if (opts.headers) {
      for (const [k, v] of Object.entries(opts.headers)) headers[k.toLowerCase()] = v;
    }

    const init: RequestInit = {
      method: opts.method ?? "GET",
      headers,
    };
    if (opts.body !== undefined) {
      init.body = JSON.stringify(opts.body);
    }

    this.logger.debug("realmid: outbound request", {
      method: init.method,
      path: opts.path,
      bearer: bearer ? redactCredential(bearer) : "<none>",
    });

    let resp: Response;
    try {
      resp = await this.fetch(url, init);
    } catch (e) {
      this.logger.error("realmid: network error", {
        method: init.method,
        path: opts.path,
        message: (e as Error).message,
      });
      throw new RealmError({
        code: "network",
        message: `network error calling ${opts.method ?? "GET"} ${opts.path}: ${(e as Error).message}`,
        cause: e,
      });
    }

    if (resp.status === 204) {
      this.logger.debug("realmid: response 204", { path: opts.path });
      return undefined as T;
    }

    const text = await resp.text();
    let body: unknown;
    if (text.length === 0) {
      body = undefined;
    } else {
      try {
        body = JSON.parse(text);
      } catch {
        body = text;
      }
    }

    if (!resp.ok) {
      const err = mapErrorResponse(resp.status, body, opts.method ?? "GET", opts.path);
      this.logger.error("realmid: request failed", {
        method: init.method,
        path: opts.path,
        status: resp.status,
        code: err.code,
      });
      throw err;
    }
    this.logger.debug("realmid: response ok", {
      method: init.method,
      path: opts.path,
      status: resp.status,
    });
    return body as T;
  }

  private buildUrl(path: string, query?: Record<string, string | number | undefined>): string {
    let url = this.baseUrl + (path.startsWith("/") ? path : `/${path}`);
    if (query) {
      const params = new URLSearchParams();
      for (const [k, v] of Object.entries(query)) {
        if (v !== undefined && v !== null && v !== "") {
          params.set(k, String(v));
        }
      }
      const qs = params.toString();
      if (qs) url += `?${qs}`;
    }
    return url;
  }
}

function mapErrorResponse(status: number, body: unknown, method: string, path: string): RealmError {
  let code: ErrorCode = statusToCode(status);
  let message = `${method} ${path} failed with HTTP ${status}`;
  let details: Record<string, unknown> | undefined;

  if (body && typeof body === "object") {
    const obj = body as Record<string, unknown>;
    const env = obj["error"];
    if (env && typeof env === "object") {
      const envObj = env as Record<string, unknown>;
      const serverCode = envObj["code"];
      if (typeof serverCode === "string" && isKnownCode(serverCode)) code = serverCode;
      const serverMsg = envObj["message"];
      if (typeof serverMsg === "string" && serverMsg.length > 0) {
        message = serverMsg;
      }
      // Siblings of `error` on the envelope (e.g. mfa_challenge_token).
      const siblings: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(obj)) {
        if (k !== "error") siblings[k] = v;
      }
      if (Object.keys(siblings).length > 0) details = siblings;
    } else if (typeof obj["code"] === "string" && isKnownCode(obj["code"] as string)) {
      // flat envelope { code, message, ... }
      code = obj["code"] as ErrorCode;
      if (typeof obj["message"] === "string") message = obj["message"];
      const siblings: Record<string, unknown> = {};
      for (const [k, v] of Object.entries(obj)) {
        if (k !== "code" && k !== "message") siblings[k] = v;
      }
      if (Object.keys(siblings).length > 0) details = siblings;
    }
  }

  return new RealmError({ code, message, httpStatus: status, details });
}

File: go/http.go
package realmid

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// httpClient is the SDK's internal wrapper around net/http. It attaches
// auth + content-type headers, parses error envelopes per SPEC §3.2, and
// emits a debug log line per request/response.
type httpClient struct {
	baseURL string
	hc      *http.Client
	logger  *slog.Logger
}

func newHTTPClient(baseURL string, hc *http.Client, logger *slog.Logger) *httpClient {
	if hc == nil {
		hc = &http.Client{Timeout: 30 * time.Second}
	}
	return &httpClient{
		baseURL: strings.TrimRight(baseURL, "/"),
		hc:      hc,
		logger:  logger,
	}
}

// requestOptions carries everything we need to issue one call.
type requestOptions struct {
	Method  string
	Path    string
	Query   map[string]string
	Body    any
	Bearer  string            // overrides any default
	Headers map[string]string // extra headers (e.g. Origin)
}

// do executes the request and JSON-decodes the response into out (if
// non-nil). On non-2xx, returns *RealmError. On network error, returns
// *RealmError{Code: network}.
func (c *httpClient) do(ctx context.Context, opts requestOptions, out any) error {
	method := opts.Method
	if method == "" {
		method = http.MethodGet
	}
	full := c.buildURL(opts.Path, opts.Query)

	var body io.Reader
	if opts.Body != nil {
		buf, err := json.Marshal(opts.Body)
		if err != nil {
			return &RealmError{Code: ErrCodeBadRequest, Message: fmt.Sprintf("marshal body: %v", err), Cause: err}
		}
		body = bytes.NewReader(buf)
	}

	req, err := http.NewRequestWithContext(ctx, method, full, body)
	if err != nil {
		return &RealmError{Code: ErrCodeBadRequest, Message: fmt.Sprintf("build request: %v", err), Cause: err}
	}
	req.Header.Set("Accept", "application/json")
	if opts.Body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if opts.Bearer != "" {
		req.Header.Set("Authorization", "Bearer "+opts.Bearer)
	}
	// ADR-056: forward the on-behalf-of user token (X-User-Token) on the typed
	// client path too, not just Realm.Do. A token stashed via WithUserToken(ctx)
	// makes user-scoped typed calls (Tenants.*, Origins.*, Auth.*) authorize on
	// the user, not the platform principal. The platform token stays the wire
	// bearer — this is additive. (Completes ADR-056; closes a partner-reported on-behalf-of gap.)
	if ut := userTokenFrom(ctx); ut != "" {
		req.Header.Set("X-User-Token", ut)
	}
	for k, v := range opts.Headers {
		req.Header.Set(k, v)
	}

	start := time.Now()
	c.logger.Debug("realmid http request",
		slog.String("method", method),
		slog.String("path", opts.Path),
		slog.String("bearer", redactCredential(opts.Bearer)),
	)

	resp, err := c.hc.Do(req)
	if err != nil {
		c.logger.Error("realmid http network error",
			slog.String("method", method),
			slog.String("path", opts.Path),
			slog.String("err", err.Error()),
		)
		return &RealmError{
			Code:    ErrCodeNetwork,
			Message: fmt.Sprintf("network error calling %s %s: %v", method, opts.Path, err),
			Cause:   err,
		}
	}
	defer func() { _ = resp.Body.Close() }()

	c.logger.Debug("realmid http response",
		slog.String("method", method),
		slog.String("path", opts.Path),
		slog.Int("status", resp.StatusCode),
		slog.Duration("latency", time.Since(start)),
	)

	if resp.StatusCode == http.StatusNoContent {
		return nil
	}

	raw, err := io.ReadAll(io.LimitReader(resp.Body, 8<<20))
	if err != nil {
		return &RealmError{Code: ErrCodeNetwork, Message: fmt.Sprintf("read body: %v", err), Cause: err}
	}

	if resp.StatusCode >= 400 {
		return mapErrorResponse(resp.StatusCode, raw, method, opts.Path)
	}

	if out == nil || len(raw) == 0 {
		return nil
	}
	if err := json.Unmarshal(raw, out); err != nil {
		return &RealmError{Code: ErrCodeServerError, Message: fmt.Sprintf("decode response: %v", err), Cause: err}
	}
	return nil
}

func (c *httpClient) buildURL(path string, query map[string]string) string {
	if !strings.HasPrefix(path, "/") {
		path = "/" + path
	}
	full := c.baseURL + path
	if len(query) == 0 {
		return full
	}
	v := url.Values{}
	for k, val := range query {
		if val == "" {
			continue
		}
		v.Set(k, val)
	}
	if enc := v.Encode(); enc != "" {
		full += "?" + enc
	}
	return full
}

func mapErrorResponse(status int, raw []byte, method, path string) *RealmError {
	return errorFromEnvelope(status, raw, fmt.Sprintf("%s %s failed with HTTP %d", method, path, status))
}

// errorFromEnvelope is the shared envelope reader behind both the typed client
// path (mapErrorResponse, which knows the method + path it called) and the
// exported ParseErrorEnvelope (which does not, and passes a status-derived
// message instead). defaultMessage is used when the body carries none.
func errorFromEnvelope(status int, raw []byte, defaultMessage string) *RealmError {
	code := statusToCode(status)
	message := defaultMessage
	var details map[string]any

	var generic map[string]any
	if len(raw) > 0 && json.Unmarshal(raw, &generic) == nil && generic != nil {
		// Envelope: { error: { code, message }, ...siblings }
		if env, ok := generic["error"].(map[string]any); ok {
			if c, ok := env["code"].(string); ok && isKnownCode(c) {
				code = ErrorCode(c)
			}
			if m, ok := env["message"].(string); ok && m != "" {
				message = m
			}
			siblings := map[string]any{}
			for k, v := range generic {
				if k == "error" {
					continue
				}
				siblings[k] = v
			}
			// Gate payloads (revocation_token, active_sessions,
			// mfa_challenge_token, …) are nested INSIDE the error object by
			// the issuer, not alongside it. Collect those too — otherwise
			// SessionLimitModal / MfaPrompt get an empty Details map.
			for k, v := range env {
				if k == "code" || k == "message" || k == "error" {
					continue
				}
				siblings[k] = v
			}
			if len(siblings) > 0 {
				details = siblings
			}
		} else {
			// Flat envelope { code, message | error: "<msg>", ...siblings }.
			// The issuer's apiErr.Response() emits { "error": "<msg>", "code":
			// "<code>" } — `error` is a STRING, not a nested object — so the
			// specific code lives at the top level alongside it. Read it here
			// (this is the shape most /auth error paths take, incl.
			// contact_admin_required, refresh_invalid, account_suspended).
			if c, ok := generic["code"].(string); ok && isKnownCode(c) {
				code = ErrorCode(c)
			}
			if m, ok := generic["message"].(string); ok && m != "" {
				message = m
			} else if s, ok := generic["error"].(string); ok && s != "" {
				message = s
			}
			siblings := map[string]any{}
			for k, v := range generic {
				if k == "code" || k == "message" || k == "error" {
					continue
				}
				siblings[k] = v
			}
			if len(siblings) > 0 {
				details = siblings
			}
		}
	}

	return &RealmError{
		Code:       code,
		Message:    message,
		HTTPStatus: status,
		Details:    details,
	}
}
